### PR TITLE
fix RateLimiterAlgorithm#callback execution sequence

### LINKED ISSUE #1375: RedisRateLimiter#isAllowed “doFinally” execution sequence problem
After my test, I found that，RedisRateLimiter#isAllowed “doFinally” execution sequence is :

1.WebClientPlugin#execute send request

2.doFinally execution RateLimiterAlgorithm#callback

3.WebClientResponsePlugin#execute write response

I think a request needs to write response ，then execution RateLimiterAlgorithm#callback, because

Concurrent requests limiter：

“Because Redis is so fast, doing the naive thing works. Just add a random token to a set at the start of a request and remove it from the set when you're done. If the set is too large, reject the request.”

from https://gist.github.com/ptarjan/e38f45f2dfe601419ca3af937fff574d#file-1-check_request_rate_limiter-rb-L11-L34

-------------------------------------------------------
so the doFinally It should be written in RateLimiterPlugin#doExecute，If so, then execution sequence is :

1.WebClientPlugin#execute send request

2.WebClientResponsePlugin#execute write response

3.doFinally execution RateLimiterAlgorithm#callback

### PR DESCRIPTION
**Pull request by WYUSIG**
fixes #1375 move the RateLimiterAlgorithm callback action to RateLimiterPlugin#doExecute doFinally.

### CODE REVIEW

**Comment by WYUSIG** [comment]
Is the code not good enough? Can you tell me, I can do better.

**Review by dengliming — DISMISSED** [review]

**Comment by dengliming** [comment]
@WYUSIG Thanks for contribution. I took a quick glance at the changes and looks good to me. Let's wait for more reviews from other committers. : )

**Review by tydhot — APPROVED** [review]

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java
Comment (tydhot) [inline soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java:83]: It has nothing to do with your changes here, but I think the expression of this diary is 农not very clear.

Comment (WYUSIG) [reply soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java:83]: After my test, I found that，RedisRateLimiter#isAllowed “doFinally” execution sequence is :

1.WebClientPlugin#execute send request

2.doFinally execution RateLimiterAlgorithm#callback

3.WebClientResponsePlugin#execute write response

I think a request needs to write response ，then execution RateLimiterAlgorithm#callback, because

Concurrent requests limiter：

“Because Redis is so fast, doing the naive thing works. Just add a random token to a set at the start of a request and remove it from the set when you're done. If the set is too large, reject the request.”

from https://gist.github.com/ptarjan/e38f45f2dfe601419ca3af937fff574d#file-1-check_request_rate_limiter-rb-L11-L34

so the doFinally It should be written in RateLimiterPlugin#doExecute，If so, then execution sequence is :

1.WebClientPlugin#execute send request

2.WebClientResponsePlugin#execute write response

3.doFinally execution RateLimiterAlgorithm#callback

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java
Comment (tydhot) [inline soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java:49]: I think the key of the attribute here maybe  need a prefix like soul or ratelimiter to make the meaning clearer.

**Review by tydhot — CHANGES_REQUESTED** [review]

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/RateLimiterPlugin.java
Comment (WYUSIG) [inline soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/RateLimiterPlugin.java:85]: Do I need to change the parameters of the callback method？because I don't think scripts will be passed to the method

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java
Comment (tydhot) [reply soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java:83]: I just happened to find that the grammar of this log is not very right. Can you fix it by the way?

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java
Comment (dengliming) [inline soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java:47]: I think you misunderstood @tydhot 's meaning. That should be something like that:
`public static final String RATE_LIMITER_KEYS = "_rate_limiter_keys";`

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java
Comment (WYUSIG) [reply soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java:83]: I found that spring cloud gateway's error log is the same
```java
catch (Exception e) {
	/*
         * We don't want a hard dependency on Redis to allow traffic. Make sure to set
         * an alert so you know if this is happening too much. Stripe's observed
         * failure rate is 0.01%.
         */
	log.error("Error determining if user allowed from redis", e);
}
```
https://github.com/spring-cloud/spring-cloud-gateway/blob/main/spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/ratelimit/RedisRateLimiter.java

line 284

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/algorithm/RateLimiterAlgorithm.java
Comment (WYUSIG) [inline soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/algorithm/RateLimiterAlgorithm.java:55]: RedisScript can get by this.getScript(), so i remove it

File: soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java
Comment (WYUSIG) [reply soul-plugin/soul-plugin-ratelimiter/src/main/java/org/dromara/soul/plugin/ratelimiter/executor/RedisRateLimiter.java:50]: This code has been fixed

**Pull request closed without merging**
